src/hnsw/build.c

```diff
@@ -343,7 +351,29 @@ static void BuildIndex(
         opts.expansion_search = metadata.expansion_search;
         opts.metric_kind = metadata.metric_kind;
     } else {
-        usearch_reserve(buildstate->usearch_index, 1100000, &error);
+        BlockNumber numBlocks = RelationGetNumberOfBlocks(heap);
+        uint32_t    estimated_row_count = 0;
+        if(numBlocks > 0) {
+            // Read the first block
+            Buffer buffer = ReadBufferExtended(heap, MAIN_FORKNUM, 0, RBM_NORMAL, NULL);
+            // Lock buffer so there won't be any new writes during this operation
+            LockBuffer(buffer, BUFFER_LOCK_SHARE);
+            // This is like converting block buffer to Page struct
+            Page page = BufferGetPage(buffer);
+            // Getting the maximum tuple index on the page
+            OffsetNumber offset = PageGetMaxOffsetNumber(page);
+            // Estimating the row count in the table
+            // There can be 3 cases
+            // 1 - new data is added and numBlocks gets increased. In this case the estimation will be lower than actual
+            // number (we need to check and increase index size) 2 - the last page is not fully associated (this is most
+            // likely to happen). In this case we will have a bit higher estimated number 3 - the last page is fully
+            // associated and we get exactly the number of rows that the table has (this is very rare case I think)
```

Comment:
In case 1, do you mean new blocks are added between `usearch_reserve()` and index insertions?

And, there actually are a lot more cases no? E.g. some of the tuples in the blocks could be deleted but not vacuumed, some blocks could represent variable length TOASTed data so offset-based heuristic above would fail, etc.

Reply:
Yes actually true, for deleted and not vacuumed rows we will allocate some more memory than needed, and the TOASTed rows will be handled by the resizing logic correctly I guess